utilities/logging.cpp

```diff
@@ -8,21 +13,112 @@ namespace logging
     static LogStream instance;
     return instance;
   }
-  LogStream::LogStream() : m_outStream{ std::cout }, m_nullStream{}, m_minLogLevel{ Level::SELF_DESTRUCT }
-  {
-    m_nullStream.setstate(std::ios_base::badbit);
-  }
+  LogStream::LogStream() : m_outStream{ std::cout }, m_minLogLevel{ Level::ERR }, m_currentLogLevel{ Level::OFF }{}
 
   void LogStream::setLogLevel(Level logLevel) { 
     m_minLogLevel = logLevel; 
   }
 
-  std::ostream& LogStream::getLogStream(Level logLevel)
+  template<typename T>
+  LogStream& LogStream::operator<<(T arg)
+  {
+    if (isActive(m_currentLogLevel, m_minLogLevel))
+    {
+      m_outStream << arg;
+    }
+    return *this;
+  }
+
+  template<>
+  LogStream& LogStream::operator<<<Level>(Level lvl)
+  {
+    if (isActive(m_currentLogLevel, m_minLogLevel) && (Level::OFF == lvl))
+    {
+      m_outStream << std::endl;
+    }
+    if (lvl >= m_minLogLevel)
+    {
+      m_currentLogLevel = lvl;
+    }
+    return *this;
+  }
+  template<>
+  LogStream& LogStream::operator<<<const char*>(const char *arg)
   {
-    if (logLevel >= m_minLogLevel)
+    if (isActive(m_currentLogLevel, m_minLogLevel))
     {
-      return m_outStream;
+      m_outStream << arg;
     }
-    return m_nullStream;
+    return *this;
   }
-}
\ No newline at end of file
+  template<>
+  LogStream& LogStream::operator<<<int>(int arg)
+  {
+    if (isActive(m_currentLogLevel, m_minLogLevel))
```

Comment:
these are so tempting to suggest a macro... then again, ERROR and INFO didn't live to tell the tale...